opentelemetry/proto/metrics/v1/metrics.proto

```diff
@@ -506,6 +503,104 @@ message DoubleHistogramDataPoint {
   // (Optional) List of exemplars collected from
   // measurements that were used to form the data point
   repeated DoubleExemplar exemplars = 8;
+
+  // When bounds have a pattern, a more efficient encoding method may be used.
+  // Only one method should be defined. For performance reason, "oneof" is not used.
+  // When more than one methods are defined, the receiver will use the first one found using this order:
+  //     explicit_bounds, linear_bounds, exponential_bounds, log_linear_bounds
+  LinearBounds linear_bounds = 9;
+  ExponentialBounds exponential_bounds = 10;
+  LogLinearBounds log_linear_bounds = 11;
+
```

Comment:
new commit added to group all "bound" types together.

Reply:
Ideally, we want:
oneof {
  ExplicitBounds explicit_bounds = 8
  LinearBounds linear_bounds = 9;
  ExponentialBounds exponential_bounds = 10;
  LogLinearBounds log_linear_bounds = 11;
}
message ExplicitBounds {
   repeated double explicit_bounds = 7
}
But " repeated double explicit_bounds = 7" is already at the top level. Though we could remove it at the cost of breaking compatibility (I heard compatibility is nice to have, but not required)

The bigger question is performance. @bogdandrutu said that oneof performance is so horrible that we should avoid it if we can.

Reply:
@bogdandrutu We have a few good questions to answer here, particularly about the `oneof`. I don't think we should bother moving field 7 or breaking compatibility, though it's sort of like keeping the worst of both worlds. 😞 

One of the questions that I found troublesome when converting to or from Prometheus is that the "le" bounds contradict the `[bounds[i-1], bounds[i])` statements here. 
(I still have to catch up on the youtube video above...)

Reply:
field order is a pain. We have these options:
1. Append new fields at the end, keeping field number order, but breaking logical grouping of fields. (my 1st version)
2. Insert new fields to their logical groups, breaking field number order. (my 2nd version, responding to Bogdan's comment)
3. Group fields logically and update field number in sequential order, breaking wire backward compatibility

So no option is ideal. Option 1 and 2 have no impact once compiled, it's just for human readers. So changing between 1 and 2 is easy. We need to be careful on choosing 3. 

As for "le" vs. "ge" bounds, I remember we basically punted on it, saying "we could add an optional boolean flag later". The bigger issue is how will a message receiver handles it. If a receiver is natively "le", it is difficult to convert incoming  "ge" buckets to "le", and vise versa. The good news is that in practice, ge vs le usually has little impact on "continuous" number set, where we are concerned on the area under a histogram curve, between two bucket bounds. As long as "area" is concerned, inclusiveness/exclusiveness at a bound is irrelevant. Even for discrete data set, le/ge will only introduce errors no more than the width of a bucket (if population count is mistakenly included to a neighboring bucket, it would introduce an error no wider than the bucket width for quantile calculation). So I think we can continue to punt the le/ge question. 

I think we are close to approval on this PR. There are only minor things like field order and "oneof". @bogdandrutu I'd like to hear from you and get this PR resolved soon.

Reply:
I vote for a oneof. Option 2 is fine with me.

Reply:
Note that to keep backward compatibility, we can only have
  repeated double explicit_bounds = 7;
  oneof bound_type {
  LinearBounds linear_bounds = 9;
  ExponentialBounds exponential_bounds = 10;
  LogLinearBounds log_linear_bounds = 11;
 }

explicit_bounds is a preexisting field. Although existing single field can be included in a new oneof without breaking backward compatibility, existing "repeated" cannot. In fact, "repeated" cannot be oneof member at all. 

So with oneof, we still leave out explicit_bounds. If we don't mind breaking compatibility, we can do:
  oneof bound_type {
  ExplicitBounds explicit_bounds = 8;
  LinearBounds linear_bounds = 9;
  ExponentialBounds exponential_bounds = 10;
  LogLinearBounds log_linear_bounds = 11;
 }
 message ExplicitBounds {
  repeated double explicit_bounds = 1;
}

Reply:
I'm OK with your first suggestion, the backward-compatible one:
```
repeated double explicit_bounds = 7;
oneof bound_type {
LinearBounds linear_bounds = 9;
ExponentialBounds exponential_bounds = 10;
LogLinearBounds log_linear_bounds = 11;
}
```
If we're going to break the protocol in any way, I'd fix this however. If we decide to break the protocol because of labels vs attributes, maybe?